<img src="./logo.png" alt="Header" align="right" style="width: 100px;"/>
<center>
    <font size="+2">
    <b>Artificial Intelligence<br></b>
        <b>Project 1: 8-Puzzle</b>
    </font>
</center>


In this project you will implement the informed version of A* to solve the 8 puzzle. You will implement 2 different heuristic measures as defined in the slides and report the results. You will also develop your own heuristic measure to solve this problem. For this you will need to implement these heuristic measures in the class State. You will also have to research how to create a priority queue in python and how you can make the class `State` comparable. 

1.1 Add a variable g that will hold the cumulative backward cost

1.2 Implement the Tiles heuristic measure to compute h

1.3 Implement the Manhattan heuristic to compute h

1.4 Test both implementations and report the values for nodes visited and explored. 

1.5 Design you own heuristic measure and implement it (It has to be consistent)

In [1]:
import numpy as np
class State:
    def __init__(self,s, zero, prev=None,g=0):
        self.s=s
        self.prev=prev
        self.zero=zero
        self.hash=None
        self.g=g
        
    def __hash__(self):
        if self.hash:
            return self.hash
        s=0
        for i in range(8,-1,-1):
            s=10*s+self.s[i]
        self.hash=s
        return s
        
    #1,5,6
    #2,0,3
    #4,7,8
    
    #[1,5,6,2,0,3,4,7,8] zero=4
    def Successors(self):
        succ=[]
        s=self.s
        zero=self.zero
        if (zero-1)>=0 and ((zero-1)//3==zero//3):
            tmp=s.copy()
            tmp[zero-1],tmp[zero]=s[zero],s[zero-1]
            succ.append(State(tmp,zero-1,self,self.g+1))
        if (zero+1<9) and ((zero+1)//3==zero//3):
            tmp=s.copy()
            tmp[zero+1],tmp[zero]=s[zero],s[zero+1]
            succ.append(State(tmp,zero+1,self,self.g+1))
        if (zero-3)>=0:
            tmp=s.copy()
            tmp[zero-3],tmp[zero]=s[zero],s[zero-3]
            succ.append(State(tmp,zero-3,self,self.g+1))
        if (zero+3)<=8:
            tmp=s.copy()
            tmp[zero+3],tmp[zero]=s[zero],s[zero+3]
            succ.append(State(tmp,zero+3,self,self.g+1))
        
        return succ
    
    def IsGoal(self):
        solution=[1,2,3,4,5,6,7,8,0]
        s=self.s
        for i in range(9):
            if s[i]!=solution[i]:
                return False
        return True
    
    def __eq__(self, other):
        for i in range(9):
            if self.s[i]!=other.s[i]:
                return False
        return True
    
    def __lt__(self, other):
        if self.g + self.ImprovedH() < other.g + other.ImprovedH(): #We substitute self.ImprovedH() and other.ImprovedH() with self.TilesH()/self.ManhattanH() and other.TilesH()/other.ManhattanH() in order to use the Tiles Heuristic/Manhattan Distance instead of the Improved Heuristic
            return True
        return False

    
    def __str__(self):
        
        tmp=np.array(self.s)
        return str(tmp.reshape(3,3))
    
    
    
    def TilesH(self):
        solution=[1,2,3,4,5,6,7,8,0]
        h=0
        for i in range(9):
            if self.s[i]!=solution[i]:
                h=h+1       
        return h
    
    
    
    def ManhattanH(self):
        solution=[1,2,3,4,5,6,7,8,0]
        h=0
        for i in range(9):
            x1=i%3
            y1=i//3
            target=solution.index(self.s[i]) #7
            x2=target%3
            y2=target//3
            h=h+abs(y1-y2)+abs(x1-x2)
            
        return h
    
    def ImprovedH(self):
        solution=[1,2,3,4,5,6,7,8,0]
        h=0
        for i in range(9):
            if self.s[i]!=solution[i]:
                h=h+1
        for i in range(9):
            x1=i%3
            y1=i//3
            target=solution.index(self.s[i]) #7
            x2=target%3
            y2=target//3
            h=h+abs(y1-y2)+abs(x1-x2)
            
            
        return h
    
    
                
        
                

In [2]:
def Solution(state):
    path=[]
    length=0
    while state:
        path.append(str(state))
        state=state.prev
        length=length+1
    return ('\n'.join(path), length)

# A*

Change the following code to implement the A* algorithm. 

In [3]:
start=State([7,2,4,5,0,6,8,3,1],4)
print(start.ManhattanH())
print(start.TilesH())
visited=set()
solved=False
from collections import deque
from queue import PriorityQueue
q = PriorityQueue() 
q.put(start)
i=0
while not q.empty():
    i=i+1
    current=q.get()
    if current.IsGoal():
        path, length= Solution(current)
        print(path)
        print("Moves: "+str(length))
        solved=True
        break
    if current in visited:
        continue
    visited.add(current)
    for tmp in current.Successors():
        q.put(tmp)
    
if not solved: 
    print("No Solution")

print("Visited states: "+str(len(visited)))
print("Explored: "+str(i))

16
7
[[1 2 3]
 [4 5 6]
 [7 8 0]]
[[1 2 3]
 [4 5 0]
 [7 8 6]]
[[1 2 3]
 [4 0 5]
 [7 8 6]]
[[1 2 3]
 [0 4 5]
 [7 8 6]]
[[0 2 3]
 [1 4 5]
 [7 8 6]]
[[2 0 3]
 [1 4 5]
 [7 8 6]]
[[2 4 3]
 [1 0 5]
 [7 8 6]]
[[2 4 3]
 [0 1 5]
 [7 8 6]]
[[2 4 3]
 [7 1 5]
 [0 8 6]]
[[2 4 3]
 [7 1 5]
 [8 0 6]]
[[2 4 3]
 [7 0 5]
 [8 1 6]]
[[2 4 3]
 [7 5 0]
 [8 1 6]]
[[2 4 0]
 [7 5 3]
 [8 1 6]]
[[2 0 4]
 [7 5 3]
 [8 1 6]]
[[0 2 4]
 [7 5 3]
 [8 1 6]]
[[7 2 4]
 [0 5 3]
 [8 1 6]]
[[7 2 4]
 [5 0 3]
 [8 1 6]]
[[7 2 4]
 [5 3 0]
 [8 1 6]]
[[7 2 4]
 [5 3 6]
 [8 1 0]]
[[7 2 4]
 [5 3 6]
 [8 0 1]]
[[7 2 4]
 [5 0 6]
 [8 3 1]]
Moves: 21
Visited states: 166
Explored: 278


## Fill the values here

Nodes Visited (Tiles): 3396 

Nodes Visited (Manhattan): 465

Nodes Visited (Yours): 166

Nodes Explored (Tiles): 5651

Nodes Explored (Manhattan): 766

Nodes Explored (Yours): 278

Name 1 : Wolf Assi

Name 2 : Raoul Boulos

Name 3 : Joe Mechref

SO we know that between 2 heuristic, the one that is less relaxed is better because we do not want to go very far from the actual answer which is the actual amount of steps that we need to take in order to solve the puzzle. Now we know that the Tiles heuristics has the smalles "h" and that the Manhattan Distance has a slightly superior "h" so we decided to sum the"h" of both of them. In that case would have incremented h and got closer to the actual solution of the program.